astropy/visualization/lupton_rgb.py

```diff
@@ -47,60 +53,159 @@ def compute_intensity(image_r, image_g=None, image_b=None):
     return np.asarray(intensity, dtype=image_r.dtype)
 
 
-class Mapping:
+class AsinhLuptonStretch(BaseStretch):
+    r"""
+    A mapping for an asinh stretch, with some changes to the constants
+    relative to `~astropy.visualization.AsinhStretch`.
+
+    The stretch is given by:
+
+    .. math::
+        y = \frac{{\rm asinh}(Q x / stretch)}{frac/{\rm asinh}(frac Q)}.
+
+    Parameters
+    ----------
+    stretch : float, optional
+        Linear stretch of the image. ``stretch`` must be greater than 0.
+        Default is 5.
+
+    Q : float, optional
+        The asinh softening parameter. ``Q`` must be greater than 0.
+        Default is 8.
+
+    Notes
+    -----
+    Based on the asinh stretch presented in Lupton et al. 2004
+    (https://ui.adsabs.harvard.edu/abs/2004PASP..116..133L).
+
     """
-    Baseclass to map red, blue, green intensities into uint8 values.
+
+    def __init__(self, stretch=5, Q=8):
+        super().__init__()
+
+        if stretch < 0:
+            raise ValueError(f"Stretch must be non-negative! {stretch=}")
+        if Q < 0:
+            raise ValueError(f"Q must be non-negative! {Q=}")
+
+        # 32bit floating point machine epsilon; sys.float_info.epsilon is 64bit
+        epsilon = 1.0 / 2**23
+        if abs(Q) < epsilon:
+            Q = 0.1
+        else:
+            Qmax = 1e10
+            if Q > Qmax:
+                Q = Qmax
+
+        self.stretch = stretch
+        self.Q = Q
+
+        frac = 0.1
+        self._slope = frac / np.arcsinh(frac * Q)
+        self._soften = Q / float(stretch)
+
+    def __call__(self, values, clip=False, out=None):
+        values = _stretch_prepare(values, clip=clip, out=out)
+        np.multiply(values, self._soften, out=values)
+        np.arcsinh(values, out=values)
+        np.multiply(values, self._slope, out=values)
+        return values
+
+
+class AsinhZscaleLuptonStretch(AsinhLuptonStretch):
```

Comment:
My understanding of the use of zscale here is that this class isn't technically a stretch class, it is a combination of an interval class (zscale) and a stretch class (asinh). Is it possible to re-work the class to be a separate interval and stretch class? I'm assuming here the zscale for lupton is a bit different from the regular zscale since it includes the pedestal option. The asinh part should be the same as the main ``AsinhLuptonStretch`` class though right? So then I guess we actually just need ``AsinhLuptonStretch`` and ``ZScaleLuptonInterval``?

Another small comment is that I think maybe the classes should be e.g. ``LuptonAsinhStretch`` and ``LuptonZScaleInterval`` grammatically.